design/cookie-rewrite-design.md

```diff
@@ -0,0 +1,196 @@
+  - [Apache HTTPD](http://httpd.apache.org/docs/2.2/mod/mod_proxy.html#proxypassreversecookiedomain)
+- Address various open issues related to cookie attributes
+  - https://github.com/projectcontour/contour/issues/2965
+  - https://github.com/projectcontour/contour/issues/3142
+
+## Non Goals
+- Provide full ability to configure Contour/Envoy with arbitrary Lua code
+- Allow users to customize all aspects of an HTTP cookie (value, other attributes not mentioned below, etc.)
+
+## High-Level Design
+`HTTPProxy` will have new fields to configure cookie rewrite policy.
+Cookie attributes related to infrastructure and security considerations will be customizable.
+Other fields, cookie names, and values will not be made customizable.
+The fields we will offer as customizable are as follows:
+- Path
+- Domain
+- Secure
+- SameSite
+
+Initially, the added field for cookie rewriting will be added to the `Route` and `Service` elements of the `HTTPProxy`.
+Envoy provides configuration of Lua using the `typed_per_filter_config` on routes, virtual hosts, or weighted clusters.
+See [here](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/lua_filter#per-route-configuration) for more.
+
+Additionally, a future step may be to allow customization of the Contour/Envoy generated session affinity cookie directly in the "cookie" load balancing strategy.
+This design will initially require users to customize that via the general setting on the `HTTPProxy` `Route` element.
+We can possibly add some cookie manipulation parameters to the "cookie" load balancing strategy settings so the `X-Contour-Session-Affinity` cookie can be customized more cleanly.
+If/when we move/add the cookie-based hash to the request hash load balancing strategy we provide, we can provide cookie manipulation parameters there to customize any cookies generated by Contour/Envoy.
+See [this draft PR](https://github.com/projectcontour/contour/pull/3993) for an example of how we can move cookie-hashing to the request hash load balancing strategy.
+We can include cookie manipulation parameters in the new `CookieHashOptions` struct.
+
+## Detailed Design
+
+### HTTPProxy API Changes
+
+Below are the changes to the `HTTPProxy` `Route` and `Service` elements and accompanying types:
+
+```golang
+type Route struct {
+  ...
+
+  // +optional
+  CookieRewritePolicy []CookieRewritePolicy `json:"cookieRewritePolicy,omitempty"`
+}
+
+type Service struct {
+  ...
+
+  // +optional
+  CookieRewritePolicy []CookieRewritePolicy `json:"cookieRewritePolicy,omitempty"`
+}
+
+type CookieRewritePolicy struct {
+	// Name is the name of the cookie for which attributes will be rewritten.
+	Name string `json:"name"`
+
+	// +optional
+	PathRewrite *CookiePathRewrite `json:"pathRewrite,omitempty"`
+
+	// +optional
+	DomainRewrite *CookieDomainRewrite `json:"domainRewrite,omitempty"`
+
+	// +optional
+	Secure *bool `json:"secure,omitempty"`
```

Comment:
Is there any way we can set defaults for these, or is that too much of a behavior change? I guess we could also do that dynamically later based on if the route is a HTTPS one or not.

Reply:
the intent of making all these pointers was to not try to suggest any defaults since I'm not sure we can know valid values in kubebuilder validation/defaulting (besides simple regex validation on names/paths/domains)

i think we could do something as you suggest to set the `Secure` attribute automatically if the rewrite policy is attached to an HTTPS route, but that would have to be in the processing stage

Reply:
I don't know if we can *always* set the `Secure` attribute for an HTTPS route since that might be an incorrect assumption, conversely we can probably assume that we *never* set `Secure` for plain HTTP routes however